auth/idtoken/idtoken.go

```diff
@@ -0,0 +1,122 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package idtoken
+
+import (
+	"fmt"
+	"net/http"
+	"os"
+
+	"cloud.google.com/go/auth"
+	"cloud.google.com/go/auth/internal"
+	"cloud.google.com/go/auth/internal/internaldetect"
+	"cloud.google.com/go/compute/metadata"
+)
+
+// ComputeTokenFormat dictates the the token format when requesting an ID token
+// from the compute metadata service.
+type ComputeTokenFormat int
+
+const (
+	// ComputeTokenFormatDefault means the same as [ComputeTokenFormatFull].
+	ComputeTokenFormatDefault ComputeTokenFormat = iota
+	// ComputeTokenFormatStandard mean only standard JWT fields will be included
+	// in the token.
+	ComputeTokenFormatStandard
+	// ComputeTokenFormatFull means the token will include claims about the
+	// virtual machine instance and its project.
+	ComputeTokenFormatFull
+	// ComputeTokenFormatFullWithLicense means the same as
+	// [ComputeTokenFormatFull] with the addition of claims about licenses
+	// associated with the instance.
+	ComputeTokenFormatFullWithLicense
+)
+
+// Options for the configuration of creation of an ID token with
+// [NewTokenProvider].
+type Options struct {
+	// Audience is the `aud` field for the token, such as an API endpoint the
+	// token will grant access to. Required.
+	Audience string
+	// ComputeTokenFormat dictates the the token format when requesting an ID
+	// token from the compute metadata service. Optional.
+	ComputeTokenFormat ComputeTokenFormat
+	// CustomClaims specifies private non-standard claims for an ID token.
+	// Optional.
+	CustomClaims map[string]interface{}
+
+	// CredentialsFile overrides detection logic and sources a credential file
+	// from the provided filepath. Optional.
+	CredentialsFile string
+	// CredentialsJSON overrides detection logic and uses the JSON bytes as the
+	// source for the credential. Optional.
+	CredentialsJSON []byte
+	// Client configures the underlying client used to make network requests
+	// when fetching tokens. If provided this should be a fully authenticated
+	// client. Optional.
+	Client *http.Client
+}
+
+func (o *Options) client() *http.Client {
+	if o == nil || o.Client == nil {
+		return internal.CloneDefaultClient()
+	}
+	return o.Client
+}
+
+// NewTokenProvider creates a [cloud.google.com/go/auth.TokenProvider] that
+// returns ID tokens configured by the opts provided. The parameter
+// opts.Audience may not be empty.
+func NewTokenProvider(opts *Options) (auth.TokenProvider, error) {
+	if opts == nil {
+		return nil, fmt.Errorf("idtoken: opts must be provided")
+	}
+	if opts.Audience == "" {
+		return nil, fmt.Errorf("idtoken: must supply a non-empty audience")
+	}
+	if b := jsonBytes(opts); b != nil {
+		return tokenProviderFromBytes(b, opts)
+	}
+	if metadata.OnGCE() {
+		return computeTokenProvider(opts)
+	}
+	return nil, fmt.Errorf("idtoken: couldn't find any credentials")
+}
+
+func jsonBytes(opts *Options) []byte {
```

Comment:
Why not make this a method on `Options`? I believe the `opts != nil` check on 99 is unnecessary after the nil check on 83.

Reply:
done and updated.